packages/x-data-grid/src/components/GridSkeletonLoadingOverlay.tsx

```diff
@@ -0,0 +1,135 @@
+import * as React from 'react';
+import Skeleton from '@mui/material/Skeleton';
+
+import { styled } from '@mui/system';
+import { unstable_useForkRef as useForkRef } from '@mui/utils';
+import { useGridApiContext } from '../hooks/utils/useGridApiContext';
+import { gridColumnPositionsSelector, gridColumnsTotalWidthSelector } from '../hooks';
+import { GridColType, GridEventListener } from '../models';
+import { seededRandomNumberGenerator } from '../utils/utils';
+
+const DEFAULT_COLUMN_WIDTH_RANGE = [40, 80] as const;
+
+const COLUMN_WIDTH_RANGE_BY_TYPE: Partial<Record<GridColType, [number, number]>> = {
+  number: [40, 60],
+  string: [40, 80],
+  date: [40, 60],
+  dateTime: [60, 80],
+  singleSelect: [40, 80],
+} as const;
+
+const colWidthVar = (index: number) => `--colWidth-${index}`;
+
+const SkeletonOverlay = styled('div')({
+  display: 'grid',
+  overflow: 'hidden',
+});
+
+const SkeletonCell = styled('div')({
+  display: 'flex',
+  flexDirection: 'row',
+  alignItems: 'center',
+  borderBottom: '1px solid var(--DataGrid-rowBorderColor)',
+});
+
+const GridSkeletonLoadingOverlay = React.forwardRef<
+  HTMLDivElement,
+  React.HTMLAttributes<HTMLDivElement>
+>(function GridSkeletonLoadingOverlay(props, forwardedRef) {
+  const ref = React.useRef<HTMLDivElement>(null);
+  const handleRef = useForkRef(ref, forwardedRef);
+
+  const apiRef = useGridApiContext();
+
+  const dimensions = apiRef.current?.getRootDimensions();
+  const viewportHeight = dimensions?.viewportInnerSize.height ?? 0;
+
+  // @ts-expect-error Function signature expects to be called with parameters, but the implementation suggests otherwise
+  const rowHeight = apiRef.current.unstable_getRowHeight();
+  const skeletonRowsCount = Math.ceil(viewportHeight / rowHeight);
+
+  const totalWidth = gridColumnsTotalWidthSelector(apiRef);
+  const positions = gridColumnPositionsSelector(apiRef);
+  const inViewportCount = React.useMemo(
+    () => positions.filter((value) => value <= totalWidth).length,
+    [totalWidth, positions],
+  );
+  const columns = apiRef.current.getVisibleColumns().slice(0, inViewportCount);
+
+  const children = React.useMemo(() => {
+    // reseed random number generator to create stable lines betwen renders
+    const array: React.ReactNode[] = [];
+    const randomNumberBetween = seededRandomNumberGenerator(12345);
```

Comment:
I would refactor the utils to have a real `randomNumberBetween` function (not a generator), and I would avoid re-creating a generator here every time.

Reply:
Not sure why we need to set a seed and why we're not using just `Math.random()`. I would refactor that code to remove the pseudo-random generator and just implement a `randomNumberBetween` function in terms of `Math.random()`.

The code was added by @DanailH, if you have more context or a reason to keep the PRG let us know.

Reply:
I think I wrote this originally, the idea of was to have random line lengths, but stable across rerenders and SSR.

Reply:
Makes sense to keep it then.

I would rename it to something more intuitive though, something like `createRandomNumberGenerator`.

Reply:
yep, and maybe elaborate a bit more on the why in the comment.

Reply:
Thanks both, have renamed the function and expanded on the comment a bit https://github.com/mui/mui-x/pull/13293/commits/4373460f0851931c966c7d43825b489533edb132

Reply:
Good shout, have updated it https://github.com/mui/mui-x/pull/13293/commits/6cf0ffa6d8a5eb177d9b01cea2539a9212fa288e